# Eyesuite export cleaning

In [ ]:
import csv

rows_out = []
header = True
with open('eyesuite_export.csv', 'r') as csvfile:
    reader = csv.reader(csvfile, delimiter=";")
    i = 0
    for row in reader:
        
        if header:
            length = len(row)
            if length != 865:
                row.extend([None,]*(865 - length))            
                rows_out.append(row)
            header = False
        
        if row[4] == "G":
            length = len(row)
            if length != 865:
                row.extend([-1,]*(865 - length))

            print(len(row))
            rows_out.append(row)

with open('eyesuite_export_001_onlyG.csv', 'w') as csvfile:
    writer = csv.writer(csvfile, delimiter=',', quoting=csv.QUOTE_MINIMAL)

    for row in rows_out:
        writer.writerow(row)

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('001_eyesuite_export_onlyG.csv', delimiter=',') #, on_bad_lines='warn', names=range(1065))


/home/davide/.virtualenvs/cas-aml-m2/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (0,25,26,27,28) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [6]:
df['{TESTLOCENUMBER}'].value_counts()

73    45481
59     1279
Name: {TESTLOCENUMBER}, dtype: int64

In [7]:
df = df[df['{TESTLOCENUMBER}'] >= 59]
df.to_csv('001_eyesuite_export_onlyG_onlygt59.csv')

In [14]:
df.describe()

Patient ID  Date of birth      Eye  {EYENUMBER}  {PROGRAM}  \
count     46784.0        46784.0  46784.0      46784.0    46784.0   
mean         -1.0           -1.0     -1.0         -1.0       -1.0   
std           0.0            0.0      0.0          0.0        0.0   
min          -1.0           -1.0     -1.0         -1.0       -1.0   
25%          -1.0           -1.0     -1.0         -1.0       -1.0   
50%          -1.0           -1.0     -1.0         -1.0       -1.0   
75%          -1.0           -1.0     -1.0         -1.0       -1.0   
max          -1.0           -1.0     -1.0         -1.0       -1.0   

       {STIMULUSSIZE}  {STIMULUSDURATION}  {STRATEGY}  {BACKGROUND}  \
count         46784.0             46784.0     46784.0       46784.0   
mean             -1.0                -1.0        -1.0          -1.0   
std               0.0                 0.0         0.0           0.0   
min              -1.0                -1.0        -1.0          -1.0   
25%              -1.0                -1.0        -1.0          -1.0   
50%              -1.0                -1.0        -1.0          -1.0   
75%              -1.0                -1.0        -1.0          -1.0   
max              -1.0                -1.0        -1.0          -1.0   

       {EXAMMETHOD}  ...   {FALSENEGATIFCATCHTRIAL}  {REPETITIONS}  {SPHERE}  \
count       46784.0  ...                    46784.0        46784.0   46784.0   
mean           -1.0  ...                       -1.0           -1.0      -1.0   
std             0.0  ...                        0.0            0.0       0.0   
min            -1.0  ...                       -1.0           -1.0      -1.0   
25%            -1.0  ...                       -1.0           -1.0      -1.0   
50%            -1.0  ...                       -1.0           -1.0      -1.0   
75%            -1.0  ...                       -1.0           -1.0      -1.0   
max            -1.0  ...                       -1.0           -1.0      -1.0   

       {CYLINDER}   {AXIS}  {VISUS}  {PUPILSIZE}    {IOP}  {POSXYPH1PH2NV}  \
count     46784.0  46784.0  46784.0      46784.0  46784.0          39658.0   
mean         -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   
std           0.0      0.0      0.0          0.0      0.0              0.0   
min          -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   
25%          -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   
50%          -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   
75%          -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   
max          -1.0     -1.0     -1.0         -1.0     -1.0             -1.0   

       Unnamed: 32  
count      39658.0  
mean          -1.0  
std            0.0  
min           -1.0  
25%           -1.0  
50%           -1.0  
75%           -1.0  
max           -1.0  

[8 rows x 33 columns]

In [7]:
df = df[df[4] == 'G']

In [9]:
df.head()

0           1    2    3    4    5    6    7     8    9     ... 1055  \
2  00001775  1974.05.05   OD    0    G    3  100    1  31.4    0  ...  NaN   
3  00001775  1974.05.05   OS    1    G    3  100    1  31.4    0  ...  NaN   
4  00004065  1974.05.17   OD    0    G    3  100    1  31.4    0  ...  NaN   
5  00004065  1974.05.17   OS    1    G    3  100    1  31.4    0  ...  NaN   
6  00006378  1988.06.18   OD    0    G    3  100    1  31.4    0  ...  NaN   

  1056 1057 1058 1059 1060 1061 1062 1063 1064  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
6  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 1065 columns]

# Heyex export cleaning

In [39]:
import pandas as pd

# read the file created from heyex db matching the VFs with OCTs
df = pd.read_csv('002_heyex_export.csv')
df.dropna(subset=['dcm_path'], inplace=True)

In [40]:
df.drop(df.columns[:2], axis=1, inplace=True)

In [41]:
df.head()

patient_id sex image_laterality             testdate  \
6    00008702   F                R  2014-02-28 14:38:47   
7    00008702   F                L  2014-02-28 14:52:19   
8    00010154   F                R  2015-10-07 14:09:37   
9    00010154   F                L  2015-10-07 14:18:51   
22   00011193   M                R  2010-10-07 13:51:39   

       dcm_acquisition_date      dcm_series_date  \
6   2014-02-28 14:18:12.420  2014-02-28T15:16:44   
7   2014-02-28 14:19:18.952  2014-02-28T15:16:44   
8   2015-10-07 13:20:17.732  2015-10-07T15:19:01   
9   2015-10-07 13:23:01.629  2015-10-07T15:19:01   
22  2010-10-07 12:20:11.453  2010-10-07T14:18:59   

                                             dcm_path   rows  columns  slices  \
6   /mnt/heyex2_archiv/archive/MC18/17636/123964/1...  496.0    512.0    49.0   
7   /mnt/heyex2_archiv/archive/MC18/17636/123964/1...  496.0    512.0    49.0   
8   /mnt/heyex2_archiv/archive/MC13/12070/98702/15...  496.0    512.0    49.0   
9   /mnt/heyex2_archiv/archive/MC13/12070/98702/15...  496.0    512.0    49.0   
22  /mnt/heyex2_archiv/archive/MC56/55155/307796/5...  496.0    512.0    49.0   

         model  depth_resolution  along_resolution  across_resolution  
6   Spectralis               7.0              13.0               13.0  
7   Spectralis               7.0              13.0               13.0  
8   Spectralis               7.0              13.0               13.0  
9   Spectralis               7.0              13.0               13.0  
22  Spectralis               7.0              13.0               13.0

In [42]:
df['test_date'] = pd.to_datetime(df.testdate)
df['vf_test_day'] = df.test_date.dt.date
df['vf_test_time'] = df.test_date.dt.time

df.drop(['test_date', 'testdate'], axis=1, inplace=True)

In [43]:
# discard double VFs on the same eye on the same day, keeping the latest
idx = df.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['vf_test_time'].transform(max) == df['vf_test_time']

In [44]:
idx.value_counts()

True     38661
False      102
Name: vf_test_time, dtype: int64

In [45]:
df = df[idx]

In [47]:
df['slices'].value_counts()

49.0    31101
61.0     7560
Name: slices, dtype: int64

In [48]:
df.to_csv('003_eyesuite_export_cleaned.csv')

At this point, the file was transferred back to salomon so that pixel checksum and spacing could be extracted from actual OCTs.

TODO:
- drop_duplicates on image content hash
- find resolution and set priority for items (first 49 fully covering, then 61, then 49 not fully covering). 61 not preferred due to storage issues.
- Keep latest exam anyway, among selected


In [9]:
df_heyex = pd.read_csv('004_heyex_export_cleaned_moreinfo.csv')
df_heyex.drop(df_heyex.columns[:2], axis=1, inplace=True)

In [10]:
original_len = len(df_heyex)
df_heyex = df_heyex.sort_values('dcm_acquisition_date', ascending=False)
df_heyex.drop_duplicates('image_hash', inplace=True)
print(original_len - len(df_heyex))

18183


In [16]:
df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['dcm_path'].count().value_counts()

1    17144
2     1565
3       55
4        7
6        1
5        1
Name: dcm_path, dtype: int64

After deleting duplicates, we are left with ~20k OCTs, of which 17144 are already unique. The remaining 3000 need to be filtered further, as the VFs have more than one corresponding OCT.

158 volumes are not acceptable, i.e. 49-slice volumes with 768 or 1024 pixels. See next line

In [35]:
# 158 volumes are not acceptable, i.e. 49-slice volumes with 768 or 1024 pixels
df_heyex = df_heyex[~((df_heyex['slices'] == 49) & (df_heyex['columns'].isin([768, 1024])))]
df_heyex = df_heyex[~((df_heyex['slices'] == 61) & (df_heyex['columns'] == 512))]
df_heyex = df_heyex[~((df_heyex['slices'] == 49) & (df_heyex['columns'] == 384))]

In [39]:
len(df_heyex)

20309

In [51]:
df_heyex['grid_width'] = df_heyex['columns'] * df_heyex['pxl_spacing_y']
df_heyex['grid_height'] = df_heyex['slices'] * df_heyex['slice_thickness']

In [53]:
df_heyex.describe()

patient_id     rows       columns        slices  depth_resolution  \
count  2.030900e+04  20309.0  20309.000000  20309.000000           20309.0   
mean   9.379695e+06    496.0    572.643852     51.842681               7.0   
std    1.134731e+07      0.0    108.847291      5.102217               0.0   
min    8.702000e+03    496.0    512.000000     49.000000               7.0   
25%    3.843548e+06    496.0    512.000000     49.000000               7.0   
50%    8.240272e+06    496.0    512.000000     49.000000               7.0   
75%    1.201995e+07    496.0    512.000000     49.000000               7.0   
max    9.091309e+07    496.0    768.000000     61.000000               7.0   

       along_resolution  across_resolution  slice_thickness  pxl_spacing_x  \
count           20309.0            20309.0     20309.000000   2.030900e+04   
mean               13.0               13.0         0.123679   3.871993e-03   
std                 0.0                0.0         0.005907   1.888101e-07   
min                13.0               13.0         0.063708   3.867000e-03   
25%                13.0               13.0         0.120278   3.872000e-03   
50%                13.0               13.0         0.122755   3.872000e-03   
75%                13.0               13.0         0.126170   3.872000e-03   
max                13.0               13.0         0.173835   3.872000e-03   

       pxl_spacing_y    grid_width   grid_height  
count   20309.000000  20309.000000  20309.000000  
mean        0.011610      6.648952      6.412061  
std         0.000552      1.307488      0.704046  
min         0.006003      4.451328      3.886188  
25%         0.011290      5.816832      5.931058  
50%         0.011525      5.988352      6.104616  
75%         0.011844      6.740480      6.822760  
max         0.016298     11.790336     10.362924

We now set priorities to volumes:
- If volume has 61 slices, priority 1 (2nd choice)
- If volume has 49 slices and covers the 6mm grid in both directions, priority 0 (1st choice)
- If volume has 49 slices and doesn't cover the 6mm grid, priority 2 (3rd choice)

In [56]:
df_heyex[df_heyex['slices'] == 61]['grid_height'].sort_values()

18830     3.886188
1247      5.970497
38107     6.029789
1249      6.069317
3420      6.137332
           ...    
15523     9.711322
9080      9.839910
770       9.866872
15535     9.973805
9086     10.362924
Name: grid_height, Length: 4811, dtype: float64

In [57]:
df_heyex[df_heyex['slices'] == 61]['grid_width'].sort_values()

18830     4.610304
1247      7.052544
38107     7.073280
1249      7.130880
3420      7.253760
           ...    
15523    11.477760
770      11.623680
9080     11.629056
9086     11.775744
15535    11.790336
Name: grid_width, Length: 4811, dtype: float64

In [58]:
df_heyex.drop([18830], axis=0, inplace=True)

In [60]:
df_heyex.describe()

patient_id     rows       columns        slices  depth_resolution  \
count  2.030800e+04  20308.0  20308.000000  20308.000000           20308.0   
mean   9.379766e+06    496.0    572.634233     51.842230               7.0   
std    1.134759e+07      0.0    108.841338      5.101938               0.0   
min    8.702000e+03    496.0    512.000000     49.000000               7.0   
25%    3.843548e+06    496.0    512.000000     49.000000               7.0   
50%    8.243344e+06    496.0    512.000000     49.000000               7.0   
75%    1.202043e+07    496.0    512.000000     49.000000               7.0   
max    9.091309e+07    496.0    768.000000     61.000000               7.0   

       along_resolution  across_resolution  slice_thickness  pxl_spacing_x  \
count           20308.0            20308.0     20308.000000   2.030800e+04   
mean               13.0               13.0         0.123682   3.871993e-03   
std                 0.0                0.0         0.005892   1.888147e-07   
min                13.0               13.0         0.092614   3.867000e-03   
25%                13.0               13.0         0.120278   3.872000e-03   
50%                13.0               13.0         0.122755   3.872000e-03   
75%                13.0               13.0         0.126170   3.872000e-03   
max                13.0               13.0         0.173835   3.872000e-03   

       pxl_spacing_y    grid_width   grid_height  
count   20308.000000  20308.000000  20308.000000  
mean        0.011610      6.649053      6.412186  
std         0.000550      1.307442      0.703841  
min         0.008694      4.451328      4.538086  
25%         0.011290      5.816832      5.931058  
50%         0.011525      5.988352      6.104690  
75%         0.011844      6.740608      6.822821  
max         0.016298     11.790336     10.362924

In [76]:
df_heyex['priority'] = 1 # set all to 61-slice priority, since we have verified that they all cover the grid (5.97 mm is the minimum)

fortynine_covering = (df_heyex['slices'] == 49) & (df_heyex.grid_height >= 6) & (df_heyex.grid_width >= 6)
fortynine_not_covering = (df_heyex['slices'] == 49) & ((df_heyex.grid_height < 6) | (df_heyex.grid_width < 6))
# verified that the sum is equal to the length of the df
# assert 4810 + 10449 + 5049 == len(df_heyex)

df_heyex.loc[fortynine_covering, 'priority'] = 0
df_heyex.loc[fortynine_not_covering, 'priority'] = 2


In [77]:
df_heyex.head()

patient_id sex image_laterality     dcm_acquisition_date  \
37831    16400852   F                L  2020-12-16 15:03:24.824   
37830    16400852   F                R  2020-12-16 15:02:30.302   
38027    82109710   F                R  2020-12-16 14:36:43.642   
38028    82109710   F                L  2020-12-16 14:35:27.270   
32065    13052764   F                L  2020-12-16 14:02:38.442   

           dcm_series_date                                           dcm_path  \
37831  2020-12-16T16:01:48  /mnt/heyex_archiv2/MC78/77341/489016/8445861/0...   
37830  2020-12-16T16:01:48  /mnt/heyex_archiv2/MC78/77341/489016/8445846/0...   
38027  2020-12-16T15:34:38  /mnt/heyex_archiv2/MC79/78007/488983/8445036/0...   
38028  2020-12-16T15:34:38  /mnt/heyex_archiv2/MC79/78007/488983/8445030/0...   
32065  2020-12-16T15:00:54  /mnt/heyex_archiv2/MC47/46408/488965/8444674/0...   

        rows  columns  slices       model  ...  across_resolution  \
37831  496.0    512.0    49.0  Spectralis  ...               13.0   
37830  496.0    512.0    49.0  Spectralis  ...               13.0   
38027  496.0    512.0    49.0  Spectralis  ...               13.0   
38028  496.0    512.0    49.0  Spectralis  ...               13.0   
32065  496.0    512.0    49.0  Spectralis  ...               13.0   

       vf_test_day  vf_test_time slice_thickness pxl_spacing_x  pxl_spacing_y  \
37831   2020-12-16      16:24:41        0.133647      0.003872       0.012632   
37830   2020-12-16      16:14:25        0.128135      0.003872       0.012220   
38027   2020-12-16      15:19:32        0.118025      0.003872       0.011079   
38028   2020-12-16      15:29:01        0.118457      0.003872       0.011120   
32065   2020-12-16      14:52:43        0.124079      0.003872       0.011543   

                             image_hash  grid_width grid_height  priority  
37831  09e7f0c9b8df7c888687c77b50f12718    6.467584    6.548703         0  
37830  af3512ea0f08b8e5e2e2902e35c5514b    6.256640    6.278615         0  
38027  53d7f2d514b9d3652fb606477fe67a1f    5.672448    5.783225         2  
38028  b84d8ac1ee6b6151fefcdc375a21d797    5.693440    5.804393         2  
32065  a75ad1c56fadb7ca4517e93d4ce3c343    5.910016    6.079871         2  

[5 rows x 22 columns]

In [78]:
df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['dcm_path'].count().value_counts()

1    17218
2     1475
3       37
4        6
5        1
Name: dcm_path, dtype: int64

In [96]:
idx = df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['priority'].transform(min) == df_heyex['priority']
df_heyex = df_heyex[idx]

idx = df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['dcm_acquisition_date'].transform(max) == df_heyex['dcm_acquisition_date']
df_heyex = df_heyex[idx]

df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day']).filter(lambda x: len(x) == 2)

patient_id sex image_laterality     dcm_acquisition_date  \
3258     1206125   M                L  2014-06-02 13:08:25.024   
3259     1206125   M                L  2014-06-02 13:08:25.024   
3256     1206125   M                R  2014-06-02 13:07:13.067   
3257     1206125   M                R  2014-06-02 13:07:13.067   

          dcm_series_date                                           dcm_path  \
3258  2014-06-02T15:05:57  /mnt/heyex_archiv2/MC49/48432/364492/6426276/0...   
3259  2014-06-02T15:05:57  /mnt/heyex2_archiv/archive/MC6/5321/65890/1055...   
3256  2014-06-02T15:05:57  /mnt/heyex_archiv2/MC49/48432/364492/6426280/0...   
3257  2014-06-02T15:05:57  /mnt/heyex2_archiv/archive/MC6/5321/65890/1055...   

       rows  columns  slices       model  ...  across_resolution  vf_test_day  \
3258  496.0    512.0    49.0  Spectralis  ...               13.0   2014-06-02   
3259  496.0    512.0    49.0  Spectralis  ...               13.0   2014-06-02   
3256  496.0    512.0    49.0  Spectralis  ...               13.0   2014-06-02   
3257  496.0    512.0    49.0  Spectralis  ...               13.0   2014-06-02   

      vf_test_time slice_thickness pxl_spacing_x  pxl_spacing_y  \
3258      13:54:54        0.118001      0.003872       0.011056   
3259      13:54:54        0.118001      0.003872       0.011056   
3256      13:39:02        0.117227      0.003872       0.010918   
3257      13:39:02        0.117227      0.003872       0.010918   

                            image_hash  grid_width grid_height  priority  
3258  2309a8f52b3440ff6dea4e14e0a02448    5.660672    5.782049         2  
3259  e54bfa4ae5741e03cdb65daec07bd4dd    5.660672    5.782049         2  
3256  91a1a8abf7af68174d6bd2d61ca9cf66    5.590016    5.744123         2  
3257  27079b38b41312d26339d381dcff466a    5.590016    5.744123         2  

[4 rows x 22 columns]

In [97]:
df_heyex.drop([3258, 3256], axis=0, inplace=True)

In [98]:
df_heyex.groupby(['patient_id', 'image_laterality', 'vf_test_day'])['dcm_path'].count().value_counts()

1    18737
Name: dcm_path, dtype: int64

In [99]:
df_heyex.to_csv('005_heyex_export_final.csv')